lib/plugins/digging.js

```diff
@@ -10,81 +11,81 @@ function inject (bot) {
-      forceLook = true
-    }
-    cb = cb || noop
-    bot.lookAt(block.position.offset(0.5, 0.5, 0.5), forceLook, () => {
+    await bot.lookAt(block.position.offset(0.5, 0.5, 0.5), forceLook)
+    bot._client.write('block_dig', {
+      status: 0, // start digging
+      location: block.position,
+      face: 1 // hard coded to always dig from the top
+    })
+    const waitTime = bot.digTime(block)
+    waitTimeout = setTimeout(finishDigging, waitTime)
+    bot.targetDigBlock = block
+    bot.swingArm()
+
+    swingInterval = setInterval(() => {
+      bot.swingArm()
+    }, 350)
+
+    function finishDigging () {
+      clearInterval(swingInterval)
+      clearTimeout(waitTimeout)
+      swingInterval = null
+      waitTimeout = null
       bot._client.write('block_dig', {
-        status: 0, // start digging
-        location: block.position,
+        status: 2, // finish digging
+        location: bot.targetDigBlock.position,
         face: 1 // hard coded to always dig from the top
       })
-      const waitTime = bot.digTime(block)
-      waitTimeout = setTimeout(finishDigging, waitTime)
-      bot.targetDigBlock = block
-      bot.swingArm()
+      bot.targetDigBlock = null
+      bot.lastDigTime = performance.now()
+      bot._updateBlockState(block.position, 0)
+    }
 
-      swingInterval = setInterval(() => {
-        bot.swingArm()
-      }, 350)
-
-      const eventName = `blockUpdate:${block.position}`
-      bot.on(eventName, onBlockUpdate)
-
-      bot.stopDigging = () => {
-        if (bot.targetDigBlock === null) return
-        bot.removeListener(eventName, onBlockUpdate)
-        clearInterval(swingInterval)
-        clearTimeout(waitTimeout)
-        swingInterval = null
-        waitTimeout = null
-        bot._client.write('block_dig', {
-          status: 1, // cancel digging
-          location: bot.targetDigBlock.position,
-          face: 1 // hard coded to always dig from the top
-        })
-        const block = bot.targetDigBlock
-        bot.targetDigBlock = null
-        bot.lastDigTime = performance.now()
-        bot.emit('diggingAborted', block)
-        bot.stopDigging = noop
-        cb(new Error('Digging aborted'))
-      }
+    const eventName = `blockUpdate:${block.position}`
+    bot.on(eventName, onBlockUpdate)
 
-      function onBlockUpdate (oldBlock, newBlock) {
-        // vanilla server never actually interrupt digging, but some server send block update when you start digging
-        // so ignore block update if not air
-        if (newBlock.type !== 0) return
-        bot.removeListener(eventName, onBlockUpdate)
-        clearInterval(swingInterval)
-        clearTimeout(waitTimeout)
-        swingInterval = null
-        waitTimeout = null
-        bot.targetDigBlock = null
-        bot.lastDigTime = performance.now()
-        bot.emit('diggingCompleted', newBlock)
-        cb()
-      }
+    const [diggingDone] = await once(bot._internalEvents, `diggingDone:${block.position}`)
```

Comment:
can you explain why we need such an event whereas we didn't need one before ?

Reply:
I agree with your sentiment - please look at @wvffle's changes he requested: https://github.com/PrismarineJS/mineflayer/pull/1475#pullrequestreview-543115315 . It was actually like you suggested before: storing the promise and doing an await, but it was deemed unwanted. There is also discussion about this in the discord